### PR TITLE
bpo-30014: make poll-like selector's modify() method faster

### PR DESCRIPTION
**Pull request by giampaolo**
http://bugs.python.org/issue30014
This applies to poll(), epoll() and devpoll() selectors, and makes modify() method up to 11.4% faster under heavy loads.

### CODE REVIEW

File: Doc/whatsnew/3.7.rst [hunk 1]
```diff
@@ -169,6 +169,8 @@ Optimizations
   using the :func:`os.scandir` function.
   (Contributed by Serhiy Storchaka in :issue:`25996`.)
 
+* :meth:`selectors.DefaultSelector.modify` on Linux, OSX and Solaris is around
```
Comment (vstinner) [inline Doc/whatsnew/3.7.rst]: "DefaultSelector" depends on the platform, I would prefer to mention concrete classes: PollSelector, EpollSelector and DevpollSelector.

File: Lib/selectors.py [hunk 2]
```diff
@@ -362,6 +361,31 @@ def unregister(self, fileobj):
             self._poll.unregister(key.fd)
             return key
 
+        def modify(self, fileobj, events, data=None):
+            try:
+                key = self._fd_to_key[self._fileobj_lookup(fileobj)]
+            except KeyError:
+                raise KeyError("{!r} is not registered".format(fileobj)) from None
```
Comment (vstinner) [inline Lib/selectors.py]: nitpick: please add an empty line for readability.

File: Lib/selectors.py [hunk 3]
```diff
@@ -457,6 +481,31 @@ def select(self, timeout=None):
...
+            if events != key.events:
+                selector_events = 0
+                if events & EVENT_READ:
+                    selector_events |= select.EPOLLIN
+                if events & EVENT_WRITE:
+                    selector_events |= select.EPOLLOUT
+                try:
+                    self._epoll.modify(key.fd, selector_events)
+                except BaseException:
+                    super().unregister(fileobj)
```
Comment (vstinner) [inline Lib/selectors.py]: The doc says that modify() works as unregister+register, so basically the doc is still ok after your change.

Question: is it worth it to mention that modify() unregisters the FD on error, and maybe even write an unit test for that?

**Review by vstinner — COMMENTED** [review]
The overall change LGTM, but I have some remarks, and I would prefer to see Charles-François Natali approve the change first, since he is the maintainer of the module.

File: Doc/whatsnew/3.7.rst (hunk 1 above)
Comment (giampaolo) [reply Doc/whatsnew/3.7.rst]: +1

File: Lib/selectors.py (hunk 3 above)
Comment (giampaolo) [reply Lib/selectors.py]: Makes sense. I added a test case for that.

File: Doc/whatsnew/3.7.rst [hunk 4]
```diff
@@ -169,6 +169,9 @@ Optimizations
   using the :func:`os.scandir` function.
   (Contributed by Serhiy Storchaka in :issue:`25996`.)
 
+* :meth:`selectors.EpollSelector.modify`, :meth:`selectors.PollSelector.modify`
+  and :meth:`selectors.DevpollSelector.modify` are around 1.5x faster.
+  (Contributed by Giampaolo Rodola')
```
Comment (berkerpeksag) [inline Doc/whatsnew/3.7.rst]: (Contributed by Giampaolo Rodola' in :issue:\`30014\`.)

File: Lib/selectors.py (hunk 2 above)
Comment (berkerpeksag) [inline Lib/selectors.py]: `f"{fileobj!r} is not registered"` would make the shorter and cleaner.

File: Lib/test/test_selectors.py [hunk 5]
```diff
@@ -175,6 +175,31 @@ def test_modify(self):
...
+        # Make sure the fd is unregister()ed in case of error on
+        # modify(): http://bugs.python.org/issue30014
+        if self.SELECTOR.__name__ == 'EpollSelector':
+            patch = unittest.mock.patch('selectors.select.epoll')
+        elif self.SELECTOR.__name__ == 'PollSelector':
+            patch = unittest.mock.patch('selectors.select.poll')
+        elif self.SELECTOR.__name__ == 'DevpollSelector':
+            patch = unittest.mock.patch('selectors.select.devpoll')
+        else:
+            return unittest.skip("")
```
Comment (berkerpeksag) [inline Lib/test/test_selectors.py]: `unittest.skip()` is a decorator. Using `raise self.skipTest()` or `raise unittest.SkipTest()` would be more appropriate for this use case.

File: Misc/NEWS [hunk 6]
```diff
@@ -303,6 +303,10 @@ Extension Modules
 Library
 -------
 
+- bpo-30014: selectors.EpollSelector.modify, selectors.PollSelector.modify
+  and selectors.DevpollSelector.modify are around 1.5x faster.
+  (Contributed by Giampaolo Rodola')
```
Comment (berkerpeksag) [inline Misc/NEWS]: Contributed by -> Patch by (but you may skip this and only attribute your name in the whatsnew document since you're already a core developer :))

**Review by berkerpeksag — DISMISSED** [review]

File: Lib/selectors.py [hunk 7]
```diff
@@ -371,6 +373,33 @@ def unregister(self, fileobj):
...
+                selector_events |= self._EVENT_READ
+            if events & EVENT_WRITE:
+                selector_events |= self._EVENT_WRITE
+            try:
+                self._selector.modify(key.fd, selector_events)
+            except BaseException:
+                super().unregister(fileobj)
+                raise
+            changed = True
+        if data != key.data:
```
Comment (vstinner) [inline Lib/selectors.py]: Comparison might be expensive. Would it ok to use "is not"?

It's ok to use "!=" if "is not" wouldn't work.

**Comment by giampaolo** [comment]
OK, I addressed all comments in the PR.

**Comment by giampaolo** [comment]
@cf-natali

File: Lib/selectors.py [hunk 8]
```diff
@@ -371,6 +373,33 @@ def unregister(self, fileobj):
...
+            changed = True
+
+        if changed:
```
Comment (serhiy-storchaka) [inline Lib/selectors.py:397]: How common is the case of unchanged `events` and `data`? Wouldn't be faster (in average) always change `key` and save a time for setting and checking `changed`?

File: Lib/selectors.py [hunk 9]
```diff
@@ -371,6 +373,33 @@ def unregister(self, fileobj):
...
+        changed = False
+        if events != key.events:
+            selector_events = 0
+            if events & EVENT_READ:
+                selector_events |= self._EVENT_READ
+            if events & EVENT_WRITE:
+                selector_events |= self._EVENT_WRITE
+            try:
+                self._selector.modify(key.fd, selector_events)
+            except BaseException:
```
Comment (serhiy-storchaka) [inline Lib/selectors.py]: Why `BaseException`? `_PollLikeSelector.register()` uses `Exception`.

If `BaseException` is used intentionally, it can be omitted, the bare `except` has the same effect.

File: Doc/whatsnew/3.7.rst [hunk 10]
```diff
@@ -234,6 +234,9 @@ Optimizations
   expressions <re>`.  Searching some patterns can now be up to 20 times faster.
   (Contributed by Serhiy Storchaka in :issue:`30285`.)
 
+* :meth:`selectors.EpollSelector.modify`, :meth:`selectors.PollSelector.modify`
+  and :meth:`selectors.DevpollSelector.modify` may be up to 11.4% faster under
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.7.rst]: Such precision is unbelievable.

File: Lib/selectors.py (hunk 9 above)
Comment (giampaolo) [reply Lib/selectors.py]: Yes, indeed this should be `Exception`

File: Lib/selectors.py (hunk 8 above)
Comment (giampaolo) [reply Lib/selectors.py:397]: I'm not sure how common it is but `_replace()` is pretty time consuming so I think it's better to avoid it if we can. Checking `changed` variable is definitively faster than doing this all the time:

```
            key = key._replace(data=data)
            self._fd_to_key[key.fd] = key
```

File: Doc/whatsnew/3.7.rst (hunk 10 above)
Comment (giampaolo) [reply Doc/whatsnew/3.7.rst]: This number comes from the benchmark script provided in http://bugs.python.org/issue30581. Perhaps it's better to rephrase this as "around 10% faster"?

Comment (serhiy-storchaka) [reply Doc/whatsnew/3.7.rst]: This would look better to me.

**Review by serhiy-storchaka — COMMENTED** [review]
Technically the patch looks correct to me (but needs a Misc/NEWS entry). But I don't know whether is is worth to add additional coupling for 10% speedup. If the subclass of `_PollLikeSelector` implements `register()` or `unregister()`, it would need to reimplement also `modify()`.

**Comment by giampaolo** [comment]
I think this is good to go. Any objections?

**Comment by vstinner** [comment]
> I think this is good to go. Any objections?

I asked why not replacing expensive "if data != key.data:" with "if data is not key.data:", but I didn't get an answer. I suggest to use cheap "is not" for speed but also avoid weird bugs on strange comparison functions (like math.nan != math.nan? :-p).

**Comment by serhiy-storchaka** [comment]
I have no objections.

**Review by vstinner — CHANGES_REQUESTED** [review]
I would like to see my question answered ;-)

**Comment by giampaolo** [comment]
> I asked why not replacing expensive "if data != key.data:" with "if data is not key.data:", but I didn't get an answer.

Oh sorry, I missed your comment. Mmm... why would that be faster? I think I'm missing something.
FWIW also `_BaseSelectorImpl.modify()` method uses `!=` instead of `is not`.

**Comment by vstinner** [comment]
> Oh sorry, I missed your comment. Mmm... why would that be faster? I think I'm missing something.

If data is a 1 GB string, data1 != data2 requires to compare all bytes. If the object implements custom __eq__() method, I don't think that it's a good idea to call it here.

IHMO we should only optimize the obvious case: when the same data object is passed again (but modify is called to modify tracked events for example).

> FWIW also _BaseSelectorImpl.modify() method uses != instead of is not.

I suggest to also modify the base method in that case.

**Comment by giampaolo** [comment]
`is not` is indeed a bit faster:

```
~/svn/cpython {bpo-30014-selectors-modify-speedup|MERGING}$ ./python -m timeit "a = '1'*1024; b = '1' * 1024 + '2'" "a != b"
1000000 loops, best of 5: 315 nsec per loop
~/svn/cpython {bpo-30014-selectors-modify-speedup|MERGING}$ ./python -m timeit "a = '1'*1024; b = '1' * 1024 + '2'" "a is not b"
1000000 loops, best of 5: 297 nsec per loop
```

OK, I added this change.

**Comment by gvanrossum** [comment]
@serhiy-storchaka I no longer know this code very well -- do you really want my review in addition to all the other reviews you've already had? Is there a specific question you have for me?

**Comment by serhiy-storchaka** [comment]
@gvanrossum, nothing specific. Just you were one of the early committers to this module (with Charles-François Natali which seems not available now) and once had committed a patch related to optimization of the `modify()` method (d0786a1a5036c06770f1c958323387fd682e789e). I though that you can know something missed by other reviewers.

File: Lib/selectors.py [hunk 11]
```diff
@@ -252,15 +252,14 @@ def unregister(self, fileobj):
...
-        # TODO: Subclasses can probably optimize this even further.
         try:
             key = self._fd_to_key[self._fileobj_lookup(fileobj)]
         except KeyError:
             raise KeyError("{!r} is not registered".format(fileobj)) from None
         if events != key.events:
             self.unregister(fileobj)
             key = self.register(fileobj, events, data)
-        elif data != key.data:
+        elif data is not key.data:
```
Comment (gvanrossum) [inline Lib/selectors.py]: I disagree with this change. I've found nothing in the docs constraining the `__eq__` sematics of data. In the common cases, where data is either None or unchanged, `__eq__` typically still takes a shortcut when comparing the object to itself. So the story about comparing two 1Gb strings seems unlikely -- either it's the same string, in which case it takes a shortcut, or it's two different strings, in which case the comparison bails at the first difference. In case the data is e.g. a list, and the new data is a different list containing the same values, this "optimization" would actually be a slow-down. In general it smells superstitious. I think the code should show the intended semantics.

File: Lib/selectors.py [hunk 12]
```diff
@@ -371,6 +372,33 @@ def unregister(self, fileobj):
...
+        changed = False
+        if events != key.events:
+            selector_events = 0
+            if events & EVENT_READ:
+                selector_events |= self._EVENT_READ
+            if events & EVENT_WRITE:
+                selector_events |= self._EVENT_WRITE
+            try:
+                self._selector.modify(key.fd, selector_events)
+            except Exception:
```
Comment (gvanrossum) [inline Lib/selectors.py:390]: I wonder if it isn't time to switch this (and other except clauses in this module) to BaseException. Otherwise a KeyboardInterrupt (in particular) could leave us in a bad state. This has been a recurring problem in asyncio; we eventually are planning to address it in asyncio, and then having it not be a problem here would be a requirement. (Though I'd be okay with filing a separate issue for this instead of folding it into this PR.)

File: Lib/selectors.py [hunk 13]
```diff
@@ -371,6 +372,33 @@ def unregister(self, fileobj):
...
+                super().unregister(fileobj)
+                raise
+            changed = True
+        if data is not key.data:
```
Comment (gvanrossum) [inline Lib/selectors.py]: Again, I think != is not generally slower and closer to the promised semantics.

**Review by gvanrossum — CHANGES_REQUESTED** [review]

File: Lib/selectors.py (hunk 11 above)
Comment (giampaolo) [reply Lib/selectors.py]: Thanks for chiming in. I changed this back.

**Review by gvanrossum — APPROVED** [review]

**Comment by gvanrossum** [comment]
I'm happy now. The BaseException thing could happen another time.

**Comment by vstinner** [comment]
It's always possible to find something to enhance, but IMHO the change is now good enough and so I merged it.

Ok for != vs is not, Guido opinion matters more than mine since he is the coauthor with Charles-François Natali if I recall correctly.

**Comment by vstinner** [comment]
Thanks Giampaolo for this nice opyimization! I tried to convince one or two times a few months ago, but you was motivated than me and succeeded to implement a nice change to factorize the code ;-)

**Comment by vstinner** [comment]
> I'm happy now. The BaseException thing could happen another time.

Yeah please, it deserves its own change. When an except block ends with "raise", I prefer "except:" or "except BaseException:".

**Comment by giampaolo** [comment]
>> I'm happy now. The BaseException thing could happen another time.

> Yeah please, it deserves its own change. When an except block ends with "raise", I prefer "except:" or "except BaseException:".

http://bugs.python.org/issue30624
https://github.com/python/cpython/pull/2082
